doc/glossary.rst

```diff
@@ -0,0 +1,973 @@
+
+    rectangular
+        Data that can be represented as a matrix with :term:`samples` on the
+        first axis and a fixed, finite set of :term:`features` on the second
+        is called rectangular.
+
+        This term excludes samples with non-vectorial structure, such as text,
+        an image of arbitrary size, a time series of arbitrary length, a set of
+        vectors, etc. The purpose of a :term:`vectorizer` is to produce
+        rectangular forms of such data.
+
+    sample
+    samples
+        We usually use this term as a noun to indicate a single feature vector.
+        Elsewhere a sample is called an instance, data point, or observation.
+        ``n_samples`` indicates the number of samples in a dataset.
+
+    sample property
+    sample properties
+        TODO
+
+    scikit-learn-contrib
+        TODO
+
+    scorer
+        TODO
+        See :ref:`scoring_parameter`.
+        See also :term:`evaluation metric`.
+
+    sparse matrix
+        TODO
+
+    target
+    targets
+        TODO
+
+        Dependent variable or outcome variable.
+
+    unlabeled
+    unlabeled data
+        TODO
+
+Class APIs and Estimator Types
+==============================
+
+.. glossary::
+
+    classifier
+        TODO
+        Mention that within scikit-learn, all support multi-class
+        classification, defaulting to OvR.
+        Mention :func:`~base.is_classifier`.
+
+    clusterer
+        TODO
+
+    cross validation splitter
+    CV splitter
+        TODO
+
+    estimator
+        TODO
+
+        The core functionality of some estimators may also be available as a
+        :term:`function`.
+
+    feature extractor
+        A :term:`tranformer` which takes input where each sample is not
+        represented as an :term:`array-like` object of fixed length, and
+        produces an `array-like` object of :term:`features` for each sample
+        (and thus a 2-dimensional array-like for a set of samples).  In other
+        words, it (lossily) maps a non-rectangular data representation into
+        :term:`rectangular` data.
+
+    meta-estimator
+        TODO
+
+        Mention duck typing. Mention that duck typing of methods only works
+        after fitting. Mention lenient validation. ?Mention sample props.
+        Mention clone.
+
+    outlier detector
+        TODO
+
+    predictor
+        An :term:`estimator` which provides :term:`predict`.
+        This encompasses :term:`classifier`, :term:`regressor`,
+        :term:`outlier detector` and sometimes :term:`clusterer` (at least when
+        they are inductive).  In scikit-learn, if an estimator is not a
+        predictor, it is usually a :term:`transformer`.
+
+    regressor
+        TODO
+        Mention :func:`~base.is_regressor`.
+
+    transformer
+        TODO
+
+    vectorizer
+        See :term:`feature extractor`.
+
+    ``Xt``
+        Shorthand for transformed ``X``.
+
+There are further APIs specifically related to a small family of estimators,
+such as:
```

Comment:
Hm maybe we should have a list of all the different APIs / types of objects and functions? There's also the cross-validation and scorer APIs.

Reply:
I've added scorer, but I think we get into grey areas when we try to be exhaustive, so I've hedged with "such as".

Reply:
That makes sense. I think the scorers and cv splitters are very commonly used objects and they have a well-defined API, so I think these are very important to mention.